examples/pattern-matcher/query.scm

```diff
@@ -0,0 +1,196 @@
+;
+; query.scm -- Parallel queries (parallel pattern matching)
+;
+; QueryLink usage example.
+;
+; The QueryLink and the BindLink are both very similar; both search
+; the AtomSpace for groundings of the query pattern, and then perform
+; a re-write, based on the results. The only difference between the two
+; is that the BindLink returns a SetLink containing the results, whereas
+; the QueryLink returns a LinkValue containing the results. This makes
+; the QueryLink a bit nicer, because it does not pollute the AtomSpace
+; with nearly-useless SetLinks.
+;
+; Although both can be run in parallel (i.e. run in different threads),
+; there's almost no point to doing so for the BindLink, since you have
+; to wait for it to complete, and provide you with the resulting
+; SetLink. By contrast, the QueryLink can drop off results at a
+; "well-known location" in the AtomSpace, as they appear, so that
+; processing can happen in parallel: processing can start on some
+; results, even while others are still being found.
+;
+; This example uses an AnchorNode to establish a "well-known location",
+; a QueryLink to attache them there, and a DeleteLink to detach results
+; from the AnchorNode. The  ParallelLink is used to run multiple threads,
+; and SleepLink to slow it down enough to see what is happening.
+;
+; Taken as a whole, it demos a toy parallel processing pipeline.
+;
+
+(use-modules (opencog) (opencog exec))
+
+; -------------
+; Create three bits of "knowledge".
+(Evaluation
+	(Predicate "foobar") (List (Concept "funny") (Concept "thing")))
+(Evaluation
+	(Predicate "foobar") (List (Concept "funny") (Concept "story")))
+(Evaluation
+	(Predicate "foobar") (List (Concept "funny") (Concept "joke")))
+
+; -------------
+; Define a simple query. It looks for the funny stuff, and attaches
+; the result to an AnchorNode
+(define query
```

Comment:
Is this query going to run in separate thread?

Reply:
Yes ... the `ParallelLink` creates three threads, running that query in one of them.

Reply:
at bottom:
```
(define threads
   (Parallel
      (SequentialAnd
         (prtime "step one A")
         (True query)
         (True (Sleep (Number 4)))
... etc.
```